### ISSUE TITLE
Add Visitor for Formula Traversal and Transformation

### ISSUE THREAD

**Issue by PhilippWendler** [sosy-lab/java-smt#5]
These exist already in CPAchecker and should be moved to JavaSMT.

**Comment by PhilippWendler** [comment]
The visitor that exist in CPAchecker do not handle quantifiers, this needs to be added.

**Comment by cheshire** [comment]
Also would it be hard to change the constructor of `BooleanFormulaTransformationManager` to require a `BooleanFormulaManager` instead of `FormulaManager`? Latter offers no advantages, and with that the class can not be called from `BooleanFormulaManager` (as it does not have `FormulaManager` instance handy)

**Comment by kfriedberger** [comment]
For handling Quantifiers, we need a QuantifierFormulaManager, and instead of using two parameters BooleanFormulaManager and QuantifierFormulaManager, we can access both via FormulaManager. I prefer keeping the whole FormulaManager.

**Comment by cheshire** [comment]
Ah true, that's a good point. Should the traversal expose quantified formulas somehow then? E.g. `visitExists(bound, formula)`

**Comment by cheshire** [comment]
Question: why does the comment say that no guarantees are made on the iteration order? Surely it just traverses the whole formula using DFS?

**Comment by kfriedberger** [comment]
We have Visitors, that cache results. 
Additionally, we have no guaranteed order for the operands of `AND(op1,op2,op3)`, as the solver can simplify the term if one of the args is `true`.

So we can not guarantee DFS.

**Comment by cheshire** [comment]
I think it's partly a terminology issue. Maybe "topological" order would be better?

> that cache results

That doesn't change the traversal order, they just have a very short traversal: check cache, return.

> we have no guaranteed order for the operands of AND

I would still call it DFS for any traversal order of AND- or OR- arguments.

In the end of the day, at least to me, an "arbitrary order" means that the traversal can pick up any argument at any point from the "worklist".
This is an important difference, as in our case if the solver starts processing the node, it will finish processing its children before jumping to other nodes.
This allows state-full methods (keeping traversal state in class fields)
In general case, no such guarantee is possible.

**Comment by PhilippWendler** [comment]
For the portfolio solver (#8), even a guarantee for a topological iteration would be problematic, though.

I would think that a stateful visitor that relies on the iteration order should be avoided if possible anyway, as it seems very error prone. I don't think we even need it in CPAchecker, or do we? Visitors that extract things like variables and atoms are stateful but do not rely on iteration order, and all kinds of transformation like renaming variable are handled by the BooleanFormulaTransformationVisitor, which is also safe.

**Comment by cheshire** [comment]
> even a guarantee for a topological iteration would be problematic, though.

Would it? Just apply it to all formulas one by one.

**Comment by cheshire** [comment]
We need to stop code duplication with CPAchecker visitors...
One problem is that JavaSMT visitors require a `FormulaManager`,
while CPAchecker ones take in a `FormulaManagerView`.

We can do the following:

 - Move the methods `create*Environment` out of `FormulaManager`.
 - Make `FormulaManagerView` implement `FormulaManager`
 - Throw `UnsupportedOperation` for `getUnsafeFormulaManager` if we really want to stop users from accessing it (though I would let it through for now until we can get rid of all of its usages...)

**Comment by PhilippWendler** [comment]
We should not expose `UnsafeFormulaManager` from `FormulaManagerView`, otherwise people will start using it.

Why are the `create*Environment` methods related to this?

**Comment by PhilippWendler** [comment]
The following cases should be handled in the visitors:
- [x] quantifiers
- [ ] boolean variables
- [ ] xor and distinct
- [ ] for Z3: [Z3_OP_OEQ](http://research.microsoft.com/en-us/um/redmond/projects/z3/code/group__capi.html#ga1fe4399e5468621e2a799a680c6667cd) (should probably throw UnsupportedOperationException)

**Comment by cheshire** [comment]
@PhilippWendler 

 - We can throw an exception in `FormulaManagerView` on `getUnsafeFormulaManager` and mark it deprecated for IDE support.
 - `create*Environment` methods are related because `FormulaManager` interface implements them.
If we make `FormulaManagerView` implement `FormulaManager` (like other views do), we'll have to implement them as well.

Otherwise we are stuck with code duplication in visitors (the ones in CPAchecker want `FormulaManagerView` and the ones in `JavaSMT` want `FormulaManager`).

**Comment by cheshire** [comment]
There are a few usages of `UnsafeFormulaManager` and raw introspection methods in CPAchecker code, I don't think they can be changed to visitors nicely.

 - In the `Solver` class there is a helper method `getUnsatCore` I have used many times; It finds an UNSAT core, but first it tries to "unravel" the formula as much as it can -- eg. `AND(AND(A, B), AND(C, D))` gets turned into `AND(A, B, C, D)` and the unsat core is searched among the subsets of `A, B, C, D`. Is visitor suitable here?

```
    if (bfmgr.isAnd(constraint)) {
      for (int k = 0; k < ufmgr.getArity(constraint); k++) {
        addConstraint((BooleanFormula)ufmgr.getArg(constraint, k),
            prover, ufmgr);
```

 - In my policy iteration code I find `UF`s and manually perform Ackermannization on them --- how that can be done without `isUF` calls?

**Comment by PhilippWendler** [comment]
> In the Solver class there is a helper method getUnsatCore I have used many times; It finds an UNSAT core, but first it tries to "unravel" the formula as much as it can -- eg. AND(AND(A, B), AND(C, D)) gets turned into AND(A, B, C, D) and the unsat core is searched among the subsets of A, B, C, D. Is visitor suitable here?

Should be doable with a visitor that collects or pushes all non-And terms.
Probably this code should even be moved to JavaSMT.


> In my policy iteration code I find UFs and manually perform Ackermannization on them --- how that can be done without isUF calls?

Ackermannization should be moved to JavaSMT as well?
BooleanFormulaVisitor does not replace `isUF`, but a new `FormulaVisitor` with methods for UFs, variables, and quantifiers, should replace it.

**Comment by cheshire** [comment]
BTW why do `visitX` methods are public? Shouldn't they be protected?

**Comment by kfriedberger** [comment]
Yeah, *protected* would be nice, but then we cannot access them from BooleanFormulaManager, as it is another package.

**Comment by cheshire** [comment]
@kfriedberger yes you're right, at least `visitDefault` can be protected though.

@PhilippWendler 
> Probably this code should even be moved to JavaSMT.

It's not feasible to add as a tactic every possible transformation we can think of.
And users will think of many others.

+ For visitors which are allowed to go inside boolean atoms, would we supply the client with any way to know what operation this is?
This would be required to implement things like `splitEqualityIfPossible`.
But on the other hand the list of all possible operations is huge and not even well-defined.

**Comment by PhilippWendler** [comment]
> It's not feasible to add as a tactic every possible transformation we can think of.
> And users will think of many others.

I agree, though I would think that not this tactic but this `getUnsatCore` method will be helpful and generally useful.



> For visitors which are allowed to go inside boolean atoms, would we supply the client with any way to know what operation this is? This would be required to implement things like splitEqualityIfPossible. But on the other hand the list of all possible operations is huge and not even well-defined.

That's the big open question.

Probably the list of operators is well defined (by SMTLIB), but still too large to be useful. One option would be to have a `visitOperator(String name, List<Formula> operands)` method (with name being the SMTLIB-compliant name).
I think it would be good if we can distinguish between operators and UFs, but syntactically they look the same and I do not know whether every solver supports distinguishing between them without hard-coding a list of all operators.

Thus my current preference for this visitor interface looks like this:
```
public interface FormulaVisitor<R> {
  R visitFreeVariable(String name, FormulaType type);
  R visitBoundVariable(String name, FormulaType type);
  R visitLiteral(??? literal, FormulaType type);
  R visitUF(String name, List<Formula> args, FormulaType type);
  R visitOperator(String name, List<Formula> args, FormulaType type);
  R visitForallQuantifier(List<Formula> variables, BooleanFormula body);
  R visitExistsQuantifier(List<Formula> variables, BooleanFormula body);
}
```
This should allow to do everything the current API allows, right?

Maybe there is a valid reason to add `<T extends Formula> visitIfThenElse(BooleanFormula cond, T thenBranch, T elseBranch);`.

For recreating the same formula while traversing one would have to add a corresponding "makeOperator(String name, ...)" method though.

Suggestions for a better interface are welcome.

**Comment by cheshire** [comment]
> Probably the list of operators is well defined (by SMT-LIB)
I think the problem is that solvers have their own extensions.
Check out http://smtlib.cs.uiowa.edu/theories.shtml. It doesn't have strings or datastructures, but Z3 has both of them (though we don't support those).
The question is whether Java-SMT should support only a subset of SMT-LIB, but e.g. we support interpolation, which isn't actually defined by SMT-LIB.

> I think it would be good if we can distinguish between operators and UFs
Should we? For practical purposes, what is the difference between an operator you don't know and a UF? Applying an operator creates a stronger formula, but it has all the properties UF has.
And I think the name is "function application" and not an "operator".

> visitLiteral
Do you mean things like numerals, eg "42"?

> visitOperator
I would prefer `visitFunction`

> visitBoundVariable
Maybe should provide a pointer to the binding quantifier? Or some way to get that access?

> visitOperator(String name, List<Formula> args, FormulaType type)
How do we make sure names are consistent across different solvers?
I think typecasting between different theories can become a pain point (all the `integer-to-rational` functions).

> visitFreeVariable
On one hand this shortcut makes sense, but on the other hand a variable is just a function with no arguments.

**Comment by PhilippWendler** [comment]
Yes, with literal I mean what you call numeral. Not sure what is more common, in programming context I think it is literal. MathSAT just calls this "number".

visitFreeVariable could maybe be useful, I guess many implementors will do something different than for UFs. The default visitor could map this call to visitUF. It is also nice for symmetry to visitBoundVariable.

For visitOperator name consistency, we cannot do better than SMTLib I guess. This is essentially a replacement for `UnsafeFormulaManager.getName()`, which suffers from the same problem. I hope people will not use this method and actually check the given name, but use it for more generic things like re-creating the same operator inside a FormulaTransformationVisitor. Or should we add explicit support for this use case instead?

For UFs, I think the distinction between operators and UFs can be quite convenient, we have a lot of callers to `isUF` I think. For example visitors that apply a renaming of UFs for example should never change operators, thus it can be useful to separate them.

Yes, function application is the correct term, but if we separate UFs and operators, we cannot use it.

**Comment by cheshire** [comment]
> but use it for more generic things like re-creating the same operator inside a FormulaTransformationVisitor

How can one re-create an operator given only a String (its name)?
For now I would just do an implementation which does the same as `UnsafeFormulaManager#getName` as we really can't do much better without investing too much time...

> Yes, function application is the correct term, but if we separate UFs and operators, we cannot use it.

That's a bit pedantic, I think if there is a method `visitUF` it's obvious that "function" means "interpreted function".

**Comment by cheshire** [comment]
For a literal we can use a `String` serialization in place of `????`.
It is always fairly cheap to serialize bitvector/rational/floating point/integer to and from strings.

**Comment by cheshire** [comment]
Actually I've meant for the numeral :) 
I'm very much -1 on using the term "literal" as in logic context it has a well defined meaning: an atom or its negation.

**Comment by PhilippWendler** [comment]
> How can one re-create an operator given only a String (its name)?

I mentioned that this probably needs to be added. I do not like it but I see no other way (we have it currently in `UnsafeFormulaManager`).

> For now I would just do an implementation which does the same as UnsafeFormulaManager#getName as we really can't do much better without investing too much time...

How would you implement a visitor that for example renames all variables then? There needs to be a way to re-create the same higher-level AST node with changed arguments.

**Comment by PhilippWendler** [comment]
> For a literal we can use a String serialization in place of ????.

String can be very inconvenient if you actually want to do something with this number. Types like bitvectors and floating point have complex number formats that cannot be trivially converted into something usable and users would need to write their own parsing for this.

**Comment by cheshire** [comment]
> How would you implement a visitor that for example renames all variables then?

We can pass an extra argument `Function<List<Formula>, Formula>` which would create a function application from the provided arguments.

**Comment by PhilippWendler** [comment]
> How would you implement a visitor that for example renames all variables then?
> 
> We can pass an extra argument Function<List<Formula>, Formula> which would create a function application from the provided arguments.

Nice idea.

**Comment by cheshire** [comment]
> and users would need to write their own parsing for this

Not if we provide the corresponding classes with a convenient `of(String s)` method.
Currently we can construct a floating point number from Rational/double/BigDecimal classes.
I don't think that any of those have 1-1 mapping with SMT floats (floats represented in a given theory may be very different from the `double` in Java), and a custom class may be needed.

I _think_ the same applies to bitvectors.

**Comment by cheshire** [comment]
Coming back to the parent's comment:

>   R visitExistsQuantifier(List<Formula> variables, BooleanFormula body);

Why are these methods here? Shouldn't they be in `BooleanFormulaVisitor` and shouldn't a `FormulaVisitor` inherit from it? Otherwise, where are all the other methods for exploring the boolean structure of the formula?

**Comment by cheshire** [comment]
> Nice idea.

That might make writing constructors even more verbose than they are currently are.
Other approach would be providing constructors on request from function names through eg `FormulaManager` something like

```java
public Function<List<Formula>, Formula> getConstructorForFunction(String functionName) {};
```

Thus my version currently is:

```java

public abstract class FormulaVisitor<R> extends BooleanFormulaVisitor<R> {
  abstract R visitFreeVariable(String name, FormulaType type);
  abstract R visitBoundVariable(String name, FormulaType type, Formula boundingAST);
  abstract R visitNumeral(String numeral, FormulaType type);
  abstract R visitUF(String functionName, List<Formula> args, FormulaType type);
  abstract R visitFunctionApplication(
      String functionName,
      List<Formula> args,
      FormulaType type,
      Function<List<Formula>, Formula> newApplicationConstructor);
}
```

with constructors optionally moved to `FormulaManager`.

**Comment by PhilippWendler** [comment]
I do not understand what you mean by "constructor".

A `FormulaManager.getConstructorForFunction` method would be the same as a method that allows to create any function application by giving a string.

I did not intend to have the two visitors extend from each other. One allows to traverse the boolean part, with semantics of function applications, and the other allows to traverse all formulas, but without any semantics of function applications. This matches the use cases we have so far.

**Comment by PhilippWendler** [comment]
> visitBoundVariable
> 
> Maybe should provide a pointer to the binding quantifier? Or some way to get that access?

I do not know, but this could be quite difficult to support for some solvers.

**Comment by cheshire** [comment]
> I did not intend to have the two visitors extend from each other

If the visitor is only mean to be applied on non-boolean ASTs then it should not see any quantifiers.

But if we want it to apply it on any kind of ASTs I don't see any reason not to extend it from the boolean visitor.
The fact that you had to duplicate the functions from it suggests that we should.

**Comment by PhilippWendler** [comment]
It is certainly intended to be applicable to all kinds of ASTs. I do not see any reason for inheritance. Why should the boolean functions be handled by different methods than the non-boolean one? Either I want to iterate syntactically over the AST without caring for what functions I see, then I do not need special methods, or I care about specific operators, then I cannot traverse non-boolean formulas. Quantifiers are different because they are syntactically different (they would not be handled by `visitFunctionApplication`).

Implementations of the visitor would be considerably easier as they have a lot less methods.

**Comment by PhilippWendler** [comment]
The existence of `visitAtom` in `BooleanFormulaVisitor` would be a problem if `FormulaVisitor` inherits from `BooleanFormulaVisitor`. What would its semantics be? When would it be called?

**Comment by cheshire** [comment]
OK point taken on inheritance.

Where from should this visitor be invoked? FormulaManager itself?
I suppose not the UnsafeFormulaManager since that's what we are trying to hide.

**Comment by cheshire** [comment]
Then the interface becomes

```java
public abstract class FormulaVisitor<R> {
  abstract R visitFreeVariable(String name, FormulaType type);
  abstract R visitBoundVariable(String name, FormulaType type, Formula boundingAST);
  abstract R visitNumeral(String numeral, FormulaType type);
  abstract R visitUF(String functionName, List<Formula> args, FormulaType type);
  abstract R visitFunctionApplication(
      String functionName,
      List<Formula> args,
      FormulaType type,
      Function<List<Formula>, Formula> newApplicationConstructor);
  abstract R visitForallQuantifier(List<Formula> variables, BooleanFormula body);
  abstract R visitExistsQuantifier(List<Formula> variables, BooleanFormula body);
}
```

**Comment by PhilippWendler** [comment]
The place of the visit method depends on #13 I guess.

The `visitUF` method should maybe get an `UninterpretedFunctionDeclaration` instance (with name field added to that class?)?

**Comment by cheshire** [comment]
I've hacked together a quick call to visitor from Z3 to get a feel for what the API would look like.
I have noticed we are missing the renaming functionality, which probably can be integrated into the constructor.

The code looks like this (inside `Z3UnsafeFormulaManager`):

```java
  public FormulaType getFormulaType(Long f) {
    // TODO
    // TODO: do we support the parametrization of FormulaType?
    return null;
  }

  public <R> R visit(FormulaVisitor<R> visitor, final Long f)  {
    String name = getName(f);
    final FormulaType type = getFormulaType(f);
    List<Formula> args = new ArrayList<>();
    for (int i=0; i<getArity(f); i++) {
      // TODO: arguments might have a different typ
      long arg = getArg(f, i);
      FormulaType argumentType = getFormulaType(arg);
      args.add(getFormulaCreator().encapsulate(argumentType, arg));
    }

    if (isFreeVariable(f)) {
      return visitor.visitFreeVariable(name, type);
    } else if (isBoundVariable(f)) {

      // TODO: is it feasible to get the bounding AST?
      return visitor.visitBoundVariable(name, type, null);
    } else if (isNumber(f)) {
      return visitor.visitNumeral(ast_to_string(z3context, f), type);
    } else if (isUF(f)) {
      return visitor.visitUF(name, args, type);
    } else if (get_ast_kind(z3context, f) == Z3_QUANTIFIER_AST) {

      // TODO: get bound arguments.

      if (is_quantifier_forall(z3context, f)) {
        return visitor.visitForAllQuantifier(args,
            getFormulaCreator().encapsulateBoolean(get_quantifier_body(z3context, f))
        );
      } else {
        return visitor.visitExistsQuantifier(args,
            getFormulaCreator().encapsulateBoolean(get_quantifier_body(z3context, f))
        );
      }
    } else {

      // Any function application.
      Function<List<Formula>, Formula> constructor = new Function<List<Formula>, Formula>() {
        @Override
        public Formula apply(List<Formula> formulas) {
          return replaceArgs(getFormulaCreator().encapsulate(type, f),
              formulas);
        }
      };
      
      return visitor.visitFunctionApplication(
          name, args, type, constructor);
    }
  }
```

**Comment by cheshire** [comment]
> The visitUF method should maybe get an UninterpretedFunctionDeclaration instance

I don't know, TBH I have never used this class... I guess yes if it's easy to do so.

**Comment by cheshire** [comment]
Regarding the formula type: we don't have facilities to provide the formula type for all arguments of the function, without making the API too hard to use.
Maybe we shouldn't provide `FormulaType` at all and instead provide a function `Formula` -> `FormulaType`?

**Comment by PhilippWendler** [comment]
> Regarding the formula type: we don't have facilities to provide the formula type for all arguments of the function, without making the API too hard to use.

I do not understand? What is "the function"? What exactly is hard to use?

> Maybe we shouldn't provide FormulaType at all and instead provide a function Formula -> FormulaType?

We have that function already, but inside the visitor one does not have the current formula instance.

**Comment by cheshire** [comment]
> We have that function already, but inside the visitor one does not have the current formula instance.

Right!

Okay, here is an updated code:

```java

  public <R> R visit(FormulaVisitor<R> visitor, final Long f)  {
    String name = getName(f);
    switch (get_ast_kind(z3context, f)) {
      case Z3_NUMERAL_AST:
        return visitor.visitNumeral(
            ast_to_string(z3context, f),
            getFormulaCreator().getFormulaType(f)
        );
      case Z3_APP_AST:
        final FormulaType type = getFormulaCreator().getFormulaType(f);
        List<Formula> args = new ArrayList<>();
        for (int i=0; i<getArity(f); i++) {
          long arg = getArg(f, i);
          FormulaType argumentType = getFormulaCreator().getFormulaType(arg);
          args.add(getFormulaCreator().encapsulate(argumentType, arg));
        }

        if (isUF(f)) {
          // Special casing for UFs.
          return visitor.visitUF(name, args, type);
        } else {
          // Any function application.
          Function<List<Formula>, Formula> constructor =
              new Function<List<Formula>, Formula>() {
            @Override
            public Formula apply(List<Formula> formulas) {
              return replaceArgs(getFormulaCreator().encapsulate(type, f),
                  formulas);
            }
          };
          return visitor.visitFunctionApplication(
              name, args, type, constructor);
        }
      case Z3_VAR_AST:
        return visitor.visitFreeVariable(name,
            getFormulaCreator().getFormulaType(f));
      case Z3_QUANTIFIER_AST:
        BooleanFormula body = getFormulaCreator().encapsulateBoolean(
                  get_quantifier_body(z3context, f));
        List<Formula> qargs = new ArrayList<>();
        for (int i=0; i<get_quantifier_num_bound(z3context, f); i++) {
          long arg = get_quantifier_pattern_ast(z3context, f, i);
          FormulaType argumentType = getFormulaCreator().getFormulaType(arg);
          qargs.add(getFormulaCreator().encapsulate(argumentType, arg));
        }

        if (is_quantifier_forall(z3context, f)) {
          return visitor.visitForAllQuantifier(qargs, body);
        } else {
          return visitor.visitExistsQuantifier(qargs, body);
        }

      case Z3_SORT_AST:
      case Z3_FUNC_DECL_AST:
      case Z3_UNKNOWN_AST:
      default:
        throw new UnsupportedOperationException("Input should be a formula AST, " +
            "got unexpected type instead");
    }
  }
```

I think I'll push it after I'll add some tests.

**Comment by cheshire** [comment]
Oh right, the renaming point:

-  [ ] Should the visitor provide any renaming facilities? (to mirror `replaceArgsAndName`)?

**Comment by cheshire** [comment]
For a bound variable there definitely should be a way to identify a quantifier AST, otherwise it would not be possible to e.g. rename bound variables, as the quantifier AST which binds them is also affected by the rename.

**Comment by cheshire** [comment]
What about theory for arrays? Strings? Custom datatypes? Should we delegate all unknown cases to `visitFunctionApplication`?

**Comment by cheshire** [comment]
Also in the spirit of cutting the boilerplate, I would make the method names shorter.

 - `visitExistsQuantifier` -> `visitExists`, `visitForallQuantifier` -> `visitForAll`. I think that "exists" and "for all" are pretty obvious in the logic context and we don't need to additionally specify what those are.
 - `visitFunctionApplication` -> `visitFunction`. Again I think that is obvious that we are visiting the application, and also more consistent with `visitUF` (otherwise why one is application and one isn't?).

**Comment by cheshire** [comment]
And yet another issue:

If we talk about `FormulaTransformationVisitor` how to handle classes which would return a different output for a quantified variable? They should be either untouched or renamed altogether.

**Comment by cheshire** [comment]
> The following cases should be handled in the visitors

I would actually close the feature set on boolean visitors now.
We don't need to handle every possible case. After all, SMT atoms are not really atomic, and all kinds of "logical" things can happen inside, and if we miss a few operators it is not a problem.
Thus I'm against supporting XOR, distinct and boolean variables as special cases, as it leads to even more boilerplate.

**Comment by PhilippWendler** [comment]
> What about theory for arrays? Strings? Custom datatypes? Should we delegate all unknown cases to visitFunctionApplication?

Where is the difference between array access, string manipulation etc. to integer arithmetic functions? I do not see any, thus for me it is clear that this is a function application.

**Comment by PhilippWendler** [comment]
> The following cases should be handled in the visitors
> 
> I would actually close the feature set on boolean visitors now.
> We don't need to handle every possible case. After all, SMT atoms are not really atomic, and all kinds of "logical" things can happen inside, and if we miss a few operators it is not a problem.

So you propose to treat all `xor` terms as atoms? I disagree, it would be very surprising and inconvenient for users, especially given the fact that solvers might introduce `xor` functions by themselves.

For variables, it might be acceptable to treat them as atoms, though I think it can also be surprising, and users who would like to access these variables directly would find it much more convenient if `visitVariable` is added.

**Comment by PhilippWendler** [comment]
> If we talk about FormulaTransformationVisitor how to handle classes which would return a different output for a quantified variable? They should be either untouched or renamed altogether.

Well, we can never fully ensure that the user does not violate this because we cannot force users to use `FormulaTransformationVisitor` for formula transformations, they could always use `FormulaVisitor`. I see the following possibilities:
1. ignore
2. runtime check that method returns its argument
3. override `visitBoundVariable` method in `FormulaTransformationVisitor` and make it final
4. override `visitBoundVariable` method in `FormulaTransformationVisitor` and let it delegate to a second void method

I would probably go for 3., though it has the disadvantage that users that need to see that variable are now forced to use `FormulaVisitor` directly and re-implement the boiler-plate code.

**Comment by cheshire** [comment]
I would go for either (3) or (4), with (3) being less boilerplate and (4)
being less surprising

**Comment by PhilippWendler** [comment]
What should be done in `FormulaVisitor` for `true` and `false`?

**Comment by PhilippWendler** [comment]
Shouldn't `UnsafeFormulaManager.visit` be moved to `FormulaManager`?

**Comment by cheshire** [comment]
It's in both, one calls the other

**Comment by PhilippWendler** [comment]
For `true` and `false`, maybe we should rename `visitNumeral` to `visitConstant`? In this case I would strongly suggest to change the `String` argument to `Object`, with a clearly defined list of possible types including `Boolean` and `Number`.

**Comment by PhilippWendler** [comment]
I think that `visitFunction` is not a good name, because UFs are also functions. `visitOperator` instead?

**Comment by cheshire** [comment]
On Wed, Dec 23, 2015 at 1:32 PM, Philipp Wendler <notifications@github.com>
wrote:

> I think that visitFunction is not a good name, because UFs are also
> functions. visitOperator instead?
>
What about "and", "or" and friends? Would you call them operators?

**Comment by PhilippWendler** [comment]
Sure, boolean operator is a common term, right? I would actually always call them operators outside of SMT world instead of functions.

**Comment by cheshire** [comment]
OK, let's change to "visitOperator" then.

**Comment by cheshire** [comment]
> For true and false, maybe we should rename visitNumeral to visitConstant? In this case I would 
> strongly suggest to change the String argument to Object, with a clearly defined list of possible types 
> including Boolean and Number.

OK I'm not against it. Do you want to do the change? :P 
It's a good point that we can have all kinds of constants, e.g. strings.

**Comment by cheshire** [comment]
On FormulaVisitor: is there any reason not to give the input formula to the client? It seems weird they have to re-create it.

**Comment by PhilippWendler** [comment]
> On FormulaVisitor: is there any reason not to give the input formula to the client? It seems weird they have to re-create it.

Yes, I think if we can manage to do this in a sane way, this is a good idea and we should do this.

The main arguments against this are "yet another parameter for visit*" and the potential of confusion for implementors which argument is which, especially for methods which already have an argument of type `Formula` (like `BooleanFormulaManager.visitNot`).

So we should decide on a fixed convention (formula instance always first, or always last), and name the parameters nicely, such that IDE templating for overriding methods will pick up meaningful names that avoid confusion. Any preferences on this?

**Comment by cheshire** [comment]
But I was talking about FormulaVisitor, not BooleanFormulaVisitor.
It does not have any methods expecting a Formula parameter.

**Comment by PhilippWendler** [comment]
I thought we would to this for both visitors, if we do it?

For `FormulaVisitor`, there are the visit methods for quantifiers, which also have a `Formula` parameter.

**Comment by cheshire** [comment]
> Yes, I think if we can manage to do this in a sane way, this is a good idea and we should do this.

Another problem is that each extra argument adds to the boilerplate the client has to write in order to use the visitor.
One possible approach would be to just give clients the formula object, and then add getters Formula -> name and Formula -> FormulaType, but you really dislike the `getName` method.

**Comment by cheshire** [comment]
If we pass in a formula, could we drop the `UfDeclaration`? And re-derive at callsite if necessary?

**Comment by cheshire** [comment]
In the spirit of cutting a boilerplate, what about merging `visitExists` and `visitForall` with an extra enum value passed?

**Comment by cheshire** [comment]
Any remaining issues?

**Comment by PhilippWendler** [comment]
Well, yes, at least the ones listed here: https://github.com/sosy-lab/java-smt/issues/5#issuecomment-165364668

Also `visitQuantifier` in `BooleanFormulaVisitor` is inconsistent with the one in `FormulaVisitor`. I don't understandy why `newBodyConstructor` is necessary in the latter, if `BooleanFormulaTransformationVisitor` can do without.

What about providing `FormulaTransformationVisitor`?

I suggest to drop the existing `Recursive*Visitor`s and let callers use `(Boolean)FormulaManager.visitRecursively()`.

Should `BooleanFormulaVisitor.visit` methods also get the input formula itself? Disadvantage is the confusing parameter list, advantage is that it makes `DefaultBooleanFormulaVisitor.visitDefault()` more useful if it gets passed this value (like in 4c3ff86).

We could have methods `visitForAll`, `visitExists`, `visitUF`, and (maybe) `visitOperator` in `Default*FormulaVisitor` and let the existing methods delegate to these. This would allow users to choose which method to override, e.g., they can override `visitUF` if they only want to handle UFs and other function applications get handled by `visitDefault` automatically. They could still override the current methods, too, so no additional code is forced onto them. I think this would be nice, what do you think?

**Comment by cheshire** [comment]
Partial reply so far.

> I think this would be nice, what do you think?

I really like the code much more after the number of the methods got severely dropped, and I would like to keep it low.
The commit removing the method `visitUF` outlines the problems with it.

**Comment by PhilippWendler** [comment]
> I think this would be nice, what do you think?
> 
> I really like the code much more after the number of the methods got severely dropped, and I would like to keep it low.

I preferred it the way before. A separate method for each separate method is better readable code. You also know the general reasons against boolean parameters.

But anyway, my proposal would not force any more methods onto client code.

> The commit removing the method visitUF outlines the problems with it.

I don't see any problems solved by a9cda41e08ab835d1482418504311fa7d5f383f5?

**Comment by cheshire** [comment]
Re #5(comment): There are two problems with supporting boolean variables:

 - More boilerplate for client code.
 - `makeAtom` becomes a bad name, as variables are also atomic.

I know issues with boolean parameters, but since now it is _our_ code calling the client code, it's not such a problem (as on the client side they will always see the parameter name, the problem is usually is that the call-site is not terribly readable).

Thus to solve both of those problems I would prefer `visitAtom(isVariable)`.

Regarding `xor`: but we don't even support creating one! Thus I am against it.
(we can make a post-1.0 ticket for full XOR support, as we really can't add more features now).

Same goes for the `distinct` operation.

About `Z3_OP_OEQ`: what about it? According to the documentation it only occurs in proofs, and we do not support them so far.

**Comment by cheshire** [comment]
Also I the visitor entry point API is quite confusing right now.
Adding `visit` methods to `FormulaManager` was meant to reduce the number of entry point to two: `visit` and `visitRecursively`.
But now we still have visitors which implement their own `visit` method! Should they be removed and/or rewritten?

**Comment by PhilippWendler** [comment]
> Thus to solve both of those problems I would prefer visitAtom(isVariable).

I do not think that the amount of visitor methods that we have currently is a problem at all. Users can still use `DefaultFormulaVisitor` if they wish or make delegation calls in their own visitor.

For `visitAtom(isVariable)`, how helpful is this if you are not given the name of the variable?

> Regarding `xor`: but we don't even support creating one!

First, this is wrong.
Second, solvers may still add it to formulas on their own, thus it may exist in formulas that client code wishes to traverse over. What should happen in this case? `UnsupportedOperationException`? This also holds for `distinct`.

> About Z3_OP_OEQ: what about it? According to the documentation it only occurs in proofs, and we do not support them so far.

Great, then this should be checked and an `UnsupportedOperationException` should be thrown.

**Comment by cheshire** [comment]
> I do not think that the amount of visitor methods that we have currently is a problem at all

It is a problem, as `DefaultFormulaVisitor` loses some safety properties of non-default ones (a guarantee that the client knows about all possible cases).

> Great, then this should be checked and an UnsupportedOperationException should be thrown.

But why? `Z3_decl_kind` has a lot of different values which can never occur in an AST, should we check for all of them?

> First, this is wrong.

OK, my mistake, let's have `visitXor` if we support those.

> For visitAtom(isVariable), how helpful is this if you are not given the name of the variable?

OK let's add `visitVariable`

> What should happen in this case? 

`visitAtom` should happen.
No, I don't find it very counter-intuitive, as in the presence of integers it is already possible to encode binary predicates within atoms. E.g. `(xor A B)` can be already encoded as `(= (+ (to_int A) (to_int B)) 1)`.

**Comment by cheshire** [comment]
Also we need to fix the quantifier visitor support.

The reason I have removed the bound variables is because Princess does not (directly) expose them.
However, turns out they are required by Z3 for quantifier instantiation, hence I will be re-adding them.

For Princess we would have to go in a roundabout way. 
Since every quantifier binds exactly one variable, we would have to instantiate a visitor inside the visitor to go down and find all bound variables (using the de-Bruijn index corresponding to the current quantifier).
Getting the current index is a bit complicated, as we need the topological traversal property to increment a current index on each `visitQuantifier`.

**Comment by cheshire** [comment]
Also: what about `visitBoundVar` for *boolean* visitors? It should be different from an unbounded variable.

The collision in options between boolean and non-boolean starts to become quite annoying.

**Comment by cheshire** [comment]
And there is a whole problem of exposing standardized names in `FormulaVisitor#visitFunction`.
We need this at least to recognize non-boolean ITEs, which we've used to do with old introspection API.

This is very tricky as  we want to:
 - Gather as many standard names as we can (ITE, OR, AND, etc)
 - Recognize that we wouldn't be able to standardize across _all_ possible functions, as there are simply too many. 

Thus we can't simply return an `Enum`. One option is to return a `String`, but also provide a `String -> Enum` interface which would recognize most commonly used functions?

**Comment by PhilippWendler** [comment]
> Also: what about visitBoundVar for boolean visitors? It should be different from an unbounded variable.

Bound boolean variable? Do solvers even allow this?

> The collision in options between boolean and non-boolean starts to become quite annoying.

I don't think it's annoying, but if you do, what do you suggest doing against this?

**Comment by cheshire** [comment]
> Bound boolean variable? Do solvers even allow this?

Of course, at least Z3 does.
SMT > SAT, and in SAT quantifiers are always over boolean variables.

> I don't think it's annoying, but if you do, what do you suggest doing against this?

I have found having to synchronize the interface between the two to be quite error-prone.
If we are able to express declaration kind in a meaningful way (see the previous comment), we can make a `BooleanFormulaVisitor` to be an abstract implementation of `FormulaVisitor`.

**Comment by PhilippWendler** [comment]
> Great, then this should be checked and an UnsupportedOperationException should be thrown.
> 
> But why? Z3_decl_kind has a lot of different values which can never occur in an AST, should we check for all of them?

Yes? There have been many many cases where we made assumptions about what solvers return and we turned out to be wrong, so this clearly seems like a place where defensive checks are valuable.

>> What should happen in this case?
> 
> visitAtom should happen.
> No, I don't find it very counter-intuitive, as in the presence of integers it is already possible to encode binary predicates within atoms. E.g. `(xor A B)` can be already encoded as `(= (+ (to_int A) (to_int B)) 1)`.

So you don't find it counter-intuitive that you cannot traverse some boolean operators and boolean subtrees of formulas with `BooleanFormulaVisitor`? The fact that many SMT theories can be used to simulate boolean operators is irrelevant here I think.

**Comment by cheshire** [comment]
> Yes? There have been many many cases where we made assumptions about what solvers return and we turned out to be wrong, so this clearly seems like a place where defensive checks are valuable.

You really don't want to do that, as there are _a_lot_ of things which can never occur in AST.
In this case we have tested that, AND I am looking at Z3 source code which says that this value only occurs in proof terms.

> So you don't find it counter-intuitive that you cannot traverse some boolean operators and boolean subtrees of formulas with BooleanFormulaVisitor? The fact that many SMT theories can be used to simulate boolean operators is irrelevant here I think.

The way I think about it is this: atoms in SMT are not really "atomic".
When we are doing something only over the boolean part we are doing some over-approximation already.
By supporting more operators we can make the over-approximation less coarse, but it is always an over-approximation.
There are no invariants which can be violated by calling `visitAtom` too soon.

**Comment by cheshire** [comment]
Another important point to consider and test:
what about unary minus? Do we treat is the same way as a binary minus?
What about negative constants? Do we want `visitConstant` on them?

**Comment by cheshire** [comment]
Now that I have started working on #23, code duplication for different formula visitors becomes even larger problem. One should definitely extend the other in some way.

**Comment by cheshire** [comment]
Regarding code duplication: we can have a private redirection visitor, which would use the `FormulaVisitor` `visit` method to redirect the call to the client-supplied `BooleanFormulaVisitor`.
That can be moved back to AbstractBooleanFormulaManager`.

`visitBoolVar` is a problematic method as then `visitAtom` becomes a bad name (because, well, variable is also an atom).

I also think we should rename `visitFunction` to `visitFuncAppl` as it's the function application, not the function itself which we are visiting.

**Comment by cheshire** [comment]
On `visitBoolVar`: We can supply a `Declaration` instance from #23 to `visitAtom`, that would allow the client to get the variable name, should they desire.

**Comment by cheshire** [comment]
Code duplication is solved in https://github.com/sosy-lab/java-smt/commit/0e99dd8aed5ee31200a3019a4bf2cc9caabf284

I have also removed `visitBoolVar` in favor of using `visitAtom` and reading the attached `FuncDecl`, and added `visitBoundVar`.

**Comment by cheshire** [comment]
I am currently satisfied with the current visitor state, I think the only feature which we have disagreed upon is handling the "distinct" case separately in boolean formula visitor,
and i think we can agree it's not the killer feature (and arguments to "distinct" are not all boolean anyway).
I am closing this ticket now, the remaining item is adding visitors for formula transformation, but that could be done separately.
